build_tools/pexpect_helper.py

```diff
@@ -0,0 +1,261 @@
+        when = self.time_since_first_message()
+        self.messages.append(Message.sent(s, when))
+        return res
+
+    def sendline(self, s):
+        """ Cover over pexpect.spawn.sendline().
+            Send the given string + linesep to the tty, returning the number of bytes written.
+        """
+        return self.send(s + os.linesep)
+
+    def expect_re(self, pat, pat_desc=None, unmatched=None, **kwargs):
+        """ Cover over pexpect.spawn.expect().
+            Look through the "new" output of self.spawn until the given pattern is matched.
+            The pattern is typically a regular expression in string form, but may also be
+            any of the types accepted by pexpect.spawn.expect().
+            If the 'unmatched' parameter is given,
+            On failure, this prints an error and exits.
+        """
+        try:
+            res = self.spawn.expect(pat, **kwargs)
+            when = self.time_since_first_message()
+            self.messages.append(Message.received(self.spawn.match.group(), when))
+            return res
+        except pexpect.ExceptionPexpect as err:
+            if not pat_desc:
+                pat_desc = str(pat)
+            self.report_exception_and_exit(pat_desc, unmatched, err)
+
+    def expect_str(self, s, **kwargs):
+        """ Cover over expect_re() which accepts a literal string. """
+        return self.expect_re(re.escape(s), **kwargs)
+
+    def expect_prompt(self, *args, **kwargs):
+        """ Convenience function which matches some text and then a prompt.
+            Match the given positional arguments as expect_re, and then look
+            for a prompt, bumping the prompt counter.
+            Returns None on success, and exits on failure.
+            Example:
+               sp.sendline("echo hello world")
+               sp.expect_prompt("hello world")
+        """
+        if args:
+            self.expect_re(*args, **kwargs)
+        self.expect_re(
+            get_prompt_re(self.prompt_counter),
+            pat_desc="prompt %d" % self.prompt_counter,
+        )
+        self.prompt_counter += 1
+
+    def report_exception_and_exit(self, pat, unmatched, err):
+        """ Things have gone badly.
+            We have an exception 'err', some pexpect.ExceptionPexpect.
+            Report it to stdout, along with the offending call site.
+            If 'unmatched' is set, print it to stdout.
+        """
+        colors = self.colors()
+        if unmatched:
+            print("{BOLD}{unmatched}{RESET}".format(unmatched=unmatched, **colors))
+        if isinstance(err, pexpect.EOF):
+            msg = "EOF"
+        elif isinstance(err, pexpect.TIMEOUT):
+            msg = "TIMEOUT"
+        else:
+            msg = "UNKNOWN"
+        filename, lineno, code_context = get_callsite()
+        print("{RED}Failed to match:{NORMAL} {pat}".format(pat=escape(pat), **colors))
+        print(
+            "{msg} from {filename}:{lineno}: {code}".format(
```

Comment:
This message could follow an established error format, so editors can jump to the failing line.
`"{filename}:{lineno}: {msg} in {code}"` should work at least for vim.

That'd also need the full relative path, for example:
`tests/pexpects/history_search.py:59: TIMEOUT in sp.expect_str("5 AAC")`

Reply:
Nice, I hadn't even thought of that.